pkg/steps/multi_stage.go

```diff
@@ -0,0 +1,202 @@
+func (s *multiStageTestStep) Description() string {
+	return fmt.Sprintf("Run test %s", s.name)
+}
+
+func (s *multiStageTestStep) Requires() (ret []api.StepLink) {
+	var needsImages, needsRelease bool
+	for _, step := range append(append(s.pre, s.test...), s.post...) {
+		if s.config.IsPipelineImage(step.Image) {
+			ret = append(ret, api.InternalImageLink(api.PipelineImageStreamTagReference(step.Image)))
+		} else if s.config.BuildsImage(step.Image) {
+			needsImages = true
+		} else {
+			needsRelease = true
+		}
+	}
+	if needsImages {
+		ret = append(ret, api.ImagesReadyLink())
+	}
+	if needsRelease {
+		ret = append(ret, api.ReleaseImagesLink())
+	}
+	return
+}
+
+func (s *multiStageTestStep) Creates() []api.StepLink { return nil }
+func (s *multiStageTestStep) Provides() (api.ParameterMap, api.StepLink) {
+	return nil, nil
+}
+func (s *multiStageTestStep) SubTests() []*junit.TestCase { return s.subTests }
+
+func (s multiStageTestStep) runSteps(ctx context.Context, steps []api.TestStep, shortCircuit bool) error {
+	pods, err := s.genPods(steps)
+	if err != nil {
+		return err
+	}
+	return s.runPods(ctx, pods, shortCircuit)
+}
+
+func (s multiStageTestStep) genPods(steps []api.TestStep) ([]coreapi.Pod, error) {
+	var ret []coreapi.Pod
+	var errs []error
+	for _, step := range steps {
+		image := step.Image
+		if s.config.IsPipelineImage(step.Image) {
+			image = api.PipelineImageStream + ":" + step.Image
+		}
+		resources, err := resourcesFor(step.Resources)
+		if err != nil {
+			errs = append(errs, err)
+			continue
+		}
+		name := s.name + "-" + step.Name
+		pod, err := generateBasePod(s.jobSpec, name, step.Name, []string{"/bin/bash", "-c", "#!/bin/bash\nset -eu\n" + step.Commands}, image, resources)
+		if err != nil {
+			errs = append(errs, err)
+			continue
+		}
+		pod.Labels[multiStageTestLabel] = s.name
+		container := &pod.Spec.Containers[0]
+		container.Env = append(container.Env, []coreapi.EnvVar{
+			{Name: "NAMESPACE", Value: s.jobSpec.Namespace},
+			{Name: "JOB_NAME_SAFE", Value: strings.Replace(s.name, "_", "-", -1)},
+			{Name: "JOB_NAME_HASH", Value: s.jobSpec.JobNameHash()},
+		}...)
+		if owner := s.jobSpec.Owner(); owner != nil {
+			pod.OwnerReferences = append(pod.OwnerReferences, *owner)
+		}
+		ret = append(ret, *pod)
+	}
+	if errs != nil {
+		return nil, utilerrors.NewAggregate(errs)
+	}
+	return ret, nil
+}
+
+func (s multiStageTestStep) runPods(ctx context.Context, pods []coreapi.Pod, shortCircuit bool) error {
+	go func() {
+		<-ctx.Done()
+		log.Printf("cleanup: Deleting pods with label %s=%s", multiStageTestLabel, s.name)
+		if err := deletePods(s.podClient.Pods(s.jobSpec.Namespace), s.name); err != nil {
+			log.Printf("failed to delete pods with label %s=%s: %v", multiStageTestLabel, s.name, err)
+		}
+	}()
+	var errs []error
+	for _, pod := range pods {
+		log.Printf("Executing %q", pod.Name)
+		if _, err := createOrRestartPod(s.podClient.Pods(s.jobSpec.Namespace), &pod); err != nil {
```

Comment:
I think there is a very small opportunity for false pod name sharing. If I'm not mistaken, pod name will be `$flowname-$stepname` (https://github.com/openshift/ci-tools/pull/100/files#diff-1c5508a2984039a0cdbdce9c3650e0a7R132), so if someone is unlucky enough to have two flows like `flow=flow step=test-step` and `flow=flow-test step=step`, we could falsely attempt to share that Pod.

Reply:
Similar situations can occur today with container tests — and probably other areas where pods and other resources are created.